src/Corax/Querying/Matches/VectorSearchMatch.cs

```diff
@@ -0,0 +1,204 @@
+        while (workingBuffer.IsEmpty == false && (read = _nearestSearch.Fill(workingBuffer, _distances.GetSpace())) != 0)
+        {
+            var currentBatch = DiscardDocumentUnderSimilarity(workingBuffer.Slice(0, read), _distances.GetSpace().Slice(0, read));
+            workingBuffer = workingBuffer.Slice(currentBatch);
+            idX += currentBatch;
+        }
+
+        if (read == 0)
+        {
+            _distances.Dispose();
+            _nearestSearch.Dispose(); //no more matches
+            _returnedAllResults = true;
+        }
+
+        // Fill have to return sorted array
+        Sort.Run(matches.Slice(0, idX));
+        return idX;
+    }
+
+    private int DiscardDocumentUnderSimilarity(Span<long> matches, Span<float> distances)
+    {
+        ref var distanceRef = ref MemoryMarshal.GetReference(distances);
+        ref var matchesRef = ref MemoryMarshal.GetReference(matches);
+        int idX = 0;
+        for (var i = 0; i < matches.Length; ++i)
+        {
+            if (Unsafe.Add(ref distanceRef, i) > _maximumDistance) 
+                continue;
+            
+            Unsafe.Add(ref matchesRef, idX) = Unsafe.Add(ref matchesRef, i);
+            Unsafe.Add(ref distanceRef, idX++) = Unsafe.Add(ref distanceRef, i);
+        }
+
+        return idX;
+    }
+    
+    private void CreateDistanceBuffer(int length)
+    {
+        ref var distances = ref _distances;
+        distances.Init(_indexSearcher.Allocator, length);
+    }
+
+    private void FillAndPersistResults()
+    {
+        ref var matches = ref _matches;
+        ref var distances = ref _distances;
+        
+        matches.Init(_indexSearcher.Allocator, 128);
+        distances.Init(_indexSearcher.Allocator, 128);
+        var currentRead = 0;
+        do
+        {
+            var mBuf = matches.GetSpace();
+            var dBuf = distances.GetSpace();
+            currentRead = _nearestSearch.Fill(mBuf, dBuf);
+            var matched = DiscardDocumentUnderSimilarity(mBuf.Slice(0, currentRead), dBuf.Slice(0, currentRead));
+            matches.AddUsage(matched);
+            distances.AddUsage(matched);
+        } while (currentRead != 0);
+            
+        matches.Results.Sort(distances.Results);
+        _nearestSearch.Dispose();
+        _resultsNotPersisted = false;
+    }
+    
+    public int AndWith(Span<long> buffer, int matches)
+    {
+        if (_isEmpty)
+            return 0;
+        
+        if (_resultsNotPersisted)
+            FillAndPersistResults();
+
+        return MergeHelper.And(buffer, buffer.Slice(0, matches), _matches.Results);
+    }
+
+    public void Score(Span<long> matches, Span<float> scores, float boostFactor)
+    {
+        if (_isEmpty)
+            return;
+        
+        ref var matchesRef = ref MemoryMarshal.GetReference(matches);
+        ref var scoresRef = ref MemoryMarshal.GetReference(scores);
+        ref var distanceRef = ref MemoryMarshal.GetReference(_distances.Results);
```

Comment:
That's the idea. We have to find the positions anyway, and since incoming matches can have many documents coming from different primitives, I would say it's okay to perform a binary search. We do something similar for boosting in other primitives.

Reply:
What about when we are ordering by this score?
Won't that force us to go over the entire thing?

Reply:
Eventually, yes, but we do not have any information about whether we have other primitives or not, so it is required.